### PR TITLE
Test Builds in PRs

### PR DESCRIPTION
**Pull request by bryanculver**
# Closes: #DNE
# What's Changed

I know we recently tried to reduce the scope of our CI (see: #1628) but in trying to validate #1683 it would be nice to:
- Ensure a PR won't break build (`poetry build` run in a container already, not exhaustive, just a quick sanity check)
- Have a published image we can pull into other environments for validation (spinning up Redis+Sentinal locally is not easy)

---

This will only build and publish an image if the magic phrase "Please build a container." is left as a comment by a member of the Nautobot organization.

Psudocode of `jq` in this line: https://github.com/nautobot/nautobot/pull/1706/files#diff-755f5eaf4b3876800d7e11ae9753a8cacf8d1b6ae7ff6f18619fa4030e4d25e5R249

```
GET comments FROM this_pr
comment_matches = FOR EACH comment IN comments
    GET body, authorAssociation FROM COMMENT
    body_trigger = (body CONTAINS "Please build a container.")
    author_assoc_trigger = (authorAssociation CONTAINS "MEMBER")
    
    IF body_trigger == true AND author_assoc_trigger == true
        return true
    ELSE
        return false

ANY true IN comment_matches
```

See valid values for authorAssociation here: https://docs.github.com/en/graphql/reference/enums#commentauthorassociation

---

### Demo

See first run where no container is built: https://github.com/nautobot/nautobot/actions/runs/2241542220/attempts/1

Magic comment left: https://github.com/nautobot/nautobot/pull/1706#issuecomment-1112632857

Trigger re-run either by rerunning the entire jobs, or pick the "Check to see if we need to build a container" job and select rerun:

Screen Shot 2022-04-28 at 16 39 03

Cached (tests are simply replayed) rerun and build of container: https://github.com/nautobot/nautobot/actions/runs/2241542220/attempts/2

Resulting container: https://github.com/nautobot/nautobot/pkgs/container/nautobot-pr-validation/20899032?tag=pr-1706-py3.7

# TODO

- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- [NA] Unit, Integration Tests
- [NA] Documentation Updates (when adding/changing features)
- [NA] Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design
- [x] Wait for #1601 (will need to add build steps from that PR)

### CODE REVIEW

**Comment by nniehoff** [comment]
If we are going to build the image should we just do all of the testing in docker too?  really just thinking out loud here.  we could build the docker images once and the tests could all be some `docker run ... invoke test.xyz` and we wouldn't have to do the poetry install in EVERY job... just a thought not sure if it's worth pursuing or not

**Comment by bryanculver** [comment]
> If we are going to build the image should we just do all of the testing in docker too? really just thinking out loud here. we could build the docker images once and the tests could all be some `docker run ... invoke test.xyz` and we wouldn't have to do the poetry install in EVERY job... just a thought not sure if it's worth pursuing or not


This is hard because the example plugin isn't everywhere and is often used for testing. We really need to double down on integration tests for testing the containers.

But yes, this is the scope of #1184

**Comment by bryanculver** [comment]
Please build a container.

**Comment by bryanculver** [comment]
Please build a container.

File: .github/workflows/ci_pullrequest.yml [hunk 1]
```diff
@@ -236,3 +236,69 @@ jobs:
...
+  # Push to ghcr.io/nautobot/nautobot-dev to be able to pull down if required for validation (see: #1683)
+  container_build_check:
+    name: "Check to see if we need to build a container"
+    needs: "integration-test"
+    runs-on: "ubuntu-20.04"
+    steps:
+      - name: "Check comments for magic statement"
+        id: comment_check
+        run: |
+          echo "::set-output name=magic_comment_present::"$(gh pr view https://github.com/nautobot/nautobot/pull/${{ github.event.number }} --json comments -q "[.comments[] | {body_trigger: .body | test(\"Please build a container.\"), author_assoc_trigger: .authorAssociation | test(\"MEMBER\")} | .body_trigger == true and .author_assoc_trigger == true] | any")
```
Comment (jathanism) [inline .github/workflows/ci_pullrequest.yml:249]: IMO the magic comment should *look* like a magic comment like `[build containers]` or `!build_containers` or something like that. I'm considering aligning around how there are certain magic comments already accepted for things like GitHub commits (See: [`[skip ci]`](https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs)).

File: .github/workflows/ci_pullrequest.yml [hunk 2]
```diff
@@ -236,3 +236,69 @@ jobs:
...
+        id: "dockermeta"
+        uses: "docker/metadata-action@v3"
+        with:
+          images: "ghcr.io/nautobot/nautobot-pr-validation"
+          flavor: |
+            latest=false
+          tags: |
+            type=ref,enable=true,priority=600,prefix=pr-,suffix=-py3.7,event=pr
+          labels: |
+            org.opencontainers.image.title=Nautobot (PR VALIDATION IMAGE)
```
Comment (glennmatthews) [inline .github/workflows/ci_pullrequest.yml:288]:
```suggestion
            org.opencontainers.image.title=Nautobot (PULL REQUEST VALIDATION IMAGE)
```

**Review by glennmatthews — APPROVED** [review]

**Comment by bryanculver** [comment]
📦 ENABLE CONTAINER BUILD

**Review by glennmatthews — COMMENTED** [review]
Can you verify whether this works as intended when the PR comes from a fork, rather than just a branch within the core repository?

**Comment by jathanism** [comment]
> 👋 Thanks for opening the PR!
> 
> Friendly reminder: To have a container built of this PR a member of this organization can leave a comment saying:
> 
> > 📦 ENABLE CONTAINER BUILD
> 
> You can also disable container builds by editing or removing the comment afterwards.
> 
> Signed, 🤖 Pull Request Container Bot

I love this!

> Container built! Check it out with:
> 
> ```shell
> docker pull ghcr.io/nautobot/nautobot-pr-validation:pr-1706-py3.7
> ```
> 
> 📦🤖 Last Built: 2022-05-20 13:03:49 UTC

And this! 😍

**Review by jathanism — APPROVED** [review]

**Comment by bryanculver** [comment]
This won't have enough value as forks won't be able to publish containers either.

Will revisit another way.

**Pull request closed without merging**